xen/arch/arm/tee/ffa_msg.c

```diff
@@ -78,3 +87,48 @@ void ffa_handle_msg_send_direct_req(struct cpu_user_regs *regs, uint32_t fid)
                  resp.a4 & mask, resp.a5 & mask, resp.a6 & mask,
                  resp.a7 & mask);
 }
+
+int32_t ffa_handle_msg_send2(struct cpu_user_regs *regs)
+{
+    struct domain *src_d = current->domain;
+    struct ffa_ctx *src_ctx = src_d->arch.tee;
+    const struct ffa_part_msg_rxtx *src_msg;
+    int32_t ret;
+
+    if ( !ffa_fw_support_fid(FFA_MSG_SEND2) )
+    {
+        ret = FFA_RET_NOT_SUPPORTED;
+        goto out;
+    }
+
+    if ( !spin_trylock(&src_ctx->tx_lock) )
+    {
+        ret = FFA_RET_BUSY;
+        goto out;
+    }
+
+    src_msg = src_ctx->tx;
+
+    if ( (src_msg->send_recv_id >> 16) != ffa_get_vm_id(src_d) ||
```

Comment:
Is this check or the one below needed? From a security point of view, it adds nothing since the memory is shared with the guest so it can be changed after the check but before it's copied.

Reply:
we are recreating the header so this would not propagate and a guest trying to give a wrong ID is at least buggy or tries to impersonate so i think we should check but i need to store this in local variables to prevent modifications by the guest in flight